src/gs-update-monitor.c

```diff
@@ -500,6 +459,72 @@ schedule_update (GsUpdateMonitor *monitor,
 				       schedule_entry_scheduled_cb, helper);
 }
 
+static void
+schedule_updates_real (GsUpdateMonitor *monitor,
+		       GsAppList *apps_to_update)
+{
+	/* we have to use an updates counter to be able to know when all
+	 * updates have been scheduled */
+	monitor->num_scheduled_updates = gs_app_list_length (apps_to_update);
+	for (guint i = 0; i < monitor->num_scheduled_updates; i++) {
+		GsApp *app = gs_app_list_index (apps_to_update, i);
+		schedule_update (monitor, app,
+				 monitor->scheduled_updates_cancellable);
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
```

Comment:
You should disconnect from the `invalidated` signal here, in case the `MwscScheduler` object hangs around after you drop your reference. If it does, you might end up getting a subsequent `monitor_clear_scheduler()` callback and end up doing a double-unref.

Reply:
I connected to this using `g_signal_connect_object`, and AFAIU if we destroy GsUpdateMonitor, then the callback for this signal won't be called again. Or maybe I am wrong?

Reply:
Yes, but the code doesn’t destroy the `GsUpdateMonitor` when the `invalidated` signal is received (as far as I can tell); it only clears `monitor->scheduler`.

I realise this is a bit immaterial because you should expect the `invalidated` signal to only ever be emitted once. Fixing this would make the `GsUpdateMonitor` code more resilient to bugs in `libmogwai-schedule-client`.

Reply:
It does use `g_clear_object()`, though, so you won’t end up unreffing the scheduler more than once. My mistake.